shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Utxo.hs

```diff
@@ -210,61 +174,224 @@ genTxRetry
-              <> (fold (snd <$> wdrls)) Val.~~ deposits
-              <> refunds
+      let spendingBalance = spendingBalanceUtxo  <+> (inject $ ((sumVal (snd <$> wdrls)) <-> deposits) <+> refunds)
+          n = if (Map.size . unUTxO) utxo < (genTxStableUtxoSize defaultConstants) -- something moderate 80-120
+                 then (genTxUtxoIncrement defaultConstants)                        -- something small 2-5
+                 else 0                                                            -- no change at all
       outputAddrs <-
-        genRecipients (length inputs) keys' scripts'
+        genRecipients (length inputs + n) keys' scripts'
           >>= genPtrAddrs (_dstate dpState')
-      let mkOutputs = calcOutputsFromBalance spendingBalance outputAddrs
       -------------------------------------------------------------------------
-      -- Build a Draft Tx and use it to calculate transaction fees
+      -- Build a Draft Tx and repeatedly add to Delta until all fees are accounted for.
       -------------------------------------------------------------------------
       let draftFee = Coin 0
-          draftOutputs = snd (mkOutputs draftFee)
+          (remainderCoin,draftOutputs) = calcOutputsFromBalance spendingBalance outputAddrs draftFee
       draftTxBody <- genTxBody inputs draftOutputs certs wdrls update draftFee ttl metadataHash
       let draftTx = Tx draftTxBody (mkTxWits' draftTxBody) metadata
-          fees = minfeeBound pparams draftTx
-      -------------------------------------------------------------------------
-      -- Generate final Tx now that we have the real fees. We need to recompute
-      -- the output amounts and in turn the txBody and its witness set.
-      -------------------------------------------------------------------------
-      if spendingBalance >= fees
-        then do
-          let (actualFees', outputs') = mkOutputs fees
-              txBody = draftTxBody {_txfee = actualFees', _outputs = outputs'}
-          pure $ Right $ Tx txBody (mkTxWits' txBody) metadata
-        else do
-          let minDiffSoFar' = min minDiffSoFar (fees Val.~~ spendingBalance)
-          if n < 1
-            then
-              pure $
-                Left $
-                  GenTxExceptionInsufficientSpendingBalance
-                    InsufficientSpendingBalanceInfo
-                      { isbiDiff = minDiffSoFar'
-                      }
-            else genTxRetry (n - 1) minDiffSoFar' ge env st
+      -- We add now repeatedly add inputs until the process converges.
+      draftTx2 <- converge remainderCoin wits scripts keys' scripts' utxo pparams keySpace draftTx
+      let fee = _txfee (_body draftTx2)
+          utxoSize = (Map.size . unUTxO) utxo
+          infrequent = hash(fromIntegral (unSlotNo slot) + fromIntegral (unCoin fee) + utxoSize)::Int
+      -- infrequently we trace the final fee and the UTxO size to show progress, and to be sure the UtxO does not become empty.
```

Comment:
I would say in general we should avoid committing debug traces, even in test code.

Reply:
The algorithm has an instability in that if we don't balance  genTxStableUtxoSize  and  genTxUtxoIncrement  correctly the size of the UTxO gradually shrinks so small we cannot support generating a transaction. This writes out a bit of information about once in every 10000 calls that lets the observer watch the trend in the growth (or shrinking) of the UTxO.  This is invaluable information.  I can imagine some one changing these, or some other change that upsets this balance. I will comment this out, but not remove it.

Reply:
I think that for a long time we have all been operating under the implicit assumption that we would not check in commented code, even for debugging test code.

Reply:
I left a comment about what might go wrong. I removed the code that can test for that.